waddrmgr: optimize using each style functions

tuxcanfly opened the pull request:
refs #215 

WIP: Trying out the approach by replacing `AllAccounts` with a `each` style func first.

davecgh commented:
This looks like the discussed approach.  However, unless we plan to add the ability to allow inactive accounts, the name should probably be `ForEachAccount` instead of `ForEachActiveAccount`.

tuxcanfly commented:
Cool, let me correct the func names.

davecgh commented:
Looks good.  I know there are more coming, so I won't merge it until everything is done.

tuxcanfly commented:
Thanks, will be working on the remaining funcs.

tuxcanfly commented:
Updated the remaining funcs. Working on optimizing storage for active/inactive addresses as discussed.

tuxcanfly commented:
Refactored to store all active addresses in a bucket and use it as an index while iterating over active addresses.

Added upgrade path for  migrating from used bucket to active bucket. Running some checks, will be up for a review in a bit.

tuxcanfly commented:
To avoid confusion, we should stick to differentiating active/inactive addresses. I've replaced all references to `used` with `inactive`.

tuxcanfly commented:
OK, ready for review. @davecgh

tuxcanfly commented:
Also renamed a few methods for consistency e.g `AddrAccount` to `AddressAccount`

tuxcanfly commented:
Depends on #238 in the way we will treat used addresses and this PR optimizes in addition to each-style functions.

tuxcanfly commented:
As discussed on IRC, an address can expire after a duration and become inactive while still remaining unused. Therefore, instead of assuming that addr not present in active bucket is used, we should have an explicit db flag on address row using a flag field which determines whether the address is used or not. Similarly, the address should be marked expired after a certain configurable duration.

tuxcanfly commented:
Limiting this PR to `ForEach` impleementations, created #254 for implementing active/inactive addresses, which I'll be working on.

jrick commented on wallet/wallet.go:
```diff
@@ -693,20 +697,14 @@ func (w *Wallet) ChangePassphrase(old, new []byte) error {
+	_ = w.Manager.ForEachAccountAddress(account,
+		func(maddr waddrmgr.ManagedAddress) error {
+			used, err = maddr.Used()
+			return nil
```
This throws away the error and re-assigns the possibly true used value instead of breaking out of the loop and returning true to the caller.

jrick commented:
ActiveAccountAddresses wasn't switched over to the new style and I'd like to see some more details in each function comments about how it executes the passed function on each item, breaking early is an error is returned.  Just saying "it iterates over X" isn't descriptive enough.

tuxcanfly commented:
Moved `ActiveAccountAddresses` to `ForEachActiveAccountAddress` and updated the docs to be more descriptive. Fixed `AccountUsed` to break out and return true on first used addr.

jrick commented:
All of the calls to maybeConvertDbError will mangle the break error.  I think that errors from this package should be converted, but errors from a caller's callback should not be.

tuxcanfly commented:
Thanks, as discussed, I've added a global err `Break` to signal a early exit from the callback. Please review.

jrick commented:
ok

tuxcanfly commented:
Thanks
